### ISSUE TITLE
optimization: Improve the audio pipeline latency

### ISSUE THREAD

**Issue by ndrean** [dwyl/image-classifier#80]
The code takes a lot of care to downsize the image to improve the ML latency.

But we (I) did not treat the audio with much care. This could be a next step to improve the latency.

For example, going from stereo to mono, changing the sampling rate down to 16kHz.

Using `ffmpeg`:
- [Using ffmpeg to mono 16kHz](https://stackoverflow.com/questions/13358287/how-to-convert-any-mp3-file-to-wav-16khz-mono-16bit).
- [stereo to mono with ffmpeg](https://trac.ffmpeg.org/wiki/AudioChannelManipulation)

Perhaps Javascript offers some solutions. To explore

**Comment by LuchoTurtle** [comment]
Interesting! I wouldn't have thought of that, awesome!

I don't know the impact of reducing stereo to mono **and** reducing the sampling rate will have on the audio and if it will "lose information" before it's fed into the model but it's a really interesting topic 👀

**Comment by LuchoTurtle** [comment]
It seems like a good PR. You can open one if you want to, additions are always welcome! I think this is an adequate optimization, it will ease some load from the server :D

I'm working on a small PR to fix the deploying issues and the volume not being used (the path was wrong) - so now models will always be loaded from it without having to re-download after it's inactive. We were wrong on volumes wiping out data, they don't. :)

**Comment by ndrean** [comment]
OK. Note that I did not push further the optimisation to convert a WAV into an MP3.

This uses `lamejs` and looks like this (to be complete):

```js
import lamejs from 'lamejs';

function wavBlobToMp3Blob(wavBlob, callback) {
  const reader = new FileReader();
  reader.onload = function(e) {
    const wav = lamejs.WavHeader.readHeader(new DataView(e.target.result));
    let samples = new Int16Array(e.target.result, wav.dataOffset, wav.dataLen / 2);
    const mp3Encoder = new lamejs.Mp3Encoder(wav.channels, wav.sampleRate, 128); // 128 kbps
    const mp3Data = [];
    const sampleBlockSize = 576; // Nombre d'échantillons par canal à encoder en une fois

    for (let i = 0; i < samples.length; i += sampleBlockSize) {
      const sampleChunk = samples.subarray(i, i + sampleBlockSize);
      const mp3buf = mp3Encoder.encodeBuffer(sampleChunk);
      if (mp3buf.length > 0) {
        mp3Data.push(new Int8Array(mp3buf));
      }
    }

    const mp3buf = mp3Encoder.flush(); // Flush any remaining data
    if (mp3buf.length > 0) {
      mp3Data.push(new Int8Array(mp3buf));
    }

    const mp3Blob = new Blob(mp3Data, {type: 'audio/mp3'});
    callback(mp3Blob);
  };
  reader.readAsArrayBuffer(wavBlob);
}
```
Above, the `callback` function is as simple as  `buff => _this.upload("speech", [buff])`.

I believe this is out of scope for this application as the main idea is to suggest possible optimisation on how to reduce ML processes latency.

**Comment by LuchoTurtle** [comment]
Closed with the merge of #81

**Comment by ndrean** [comment]
I checked that the `Whisper` model wanted 16 kHz but I did not know that it seems to use `ffmpeg` under the hood to resample the audio to 16 kHz.

Check it: <https://github.com/openai/whisper/discussions/870>

In the implemented solution here, the client does it  so this saves some CPU, so it is not bad. I suggest we keep this.
